python/triton/language/semantic.py

```diff
@@ -969,16 +969,8 @@ def where(condition: tl.tensor,
         x = broadcast_impl_shape(x, condition.type.get_block_shapes(), builder)
         y = broadcast_impl_shape(y, condition.type.get_block_shapes(), builder)
 
-    # TODO: we need to check x's and y's shape?
-    x_ty = x.type.scalar
-    y_ty = y.type.scalar
-    ty = computation_type_impl(x_ty, y_ty, div_or_mod=False)
-    x = cast(x, ty, builder)
-    y = cast(y, ty, builder)
-    if x.type.is_block():
-        ret_ty = tl.block_type(ty, x.type.shape)
-    else:
-        ret_ty = ty
+    x, y = binary_op_type_checking_impl(x, y, builder, True, True)
```

Comment:
Won't we run into a problem with `tl.where(cond, ptr, int)` there?

Reply:
hmm this is a problem. I think `tl.where(cond, ptr, int)` should be invalid.

Reply:
Hmm I was looking at the C++ specs: https://docs.microsoft.com/en-us/cpp/cpp/conditional-operator-q?redirectedfrom=MSDN&view=msvc-170.

If we want to do things well, we could still allow `tl.where(cond, ptr, int)` to be valid when `int` is a constexpr that evaluated to 0.

> If both operands are of pointer types or if one is a pointer type and the other is a constant expression that evaluates to 0, pointer conversions are performed to convert them to a common type.

Reply:
I see. Makes sense

Reply:
I am trying to think of cases where we'd want `tl.where(cond, ptr, constexpr[int]=0)` and can't think of much. However. I could see potentially people wanting to do things like `tl.where(cond, ptr1, ptr2)`, where `ptr2` is actually a `None` constexpr. It is pretty common for Triton kernels to accept optional tensors, whose values are either a ptr or None. :D

Reply:
Yeah. I think this is better. At least people can explicitly cast `int` to `ptr`.